suite/tests/api/ir_riscv64.c

```diff
@@ -1104,6 +1114,10 @@ test_jump_and_branch(void *dc)
     instr = INSTR_CREATE_auipc(dc, opnd_create_reg(DR_REG_A0),
                                opnd_create_pc(pc + (3 << 12)));
     test_instr_encoding_auipc(dc, OP_auipc, pc, instr);
+
+    instr = INSTR_CREATE_auipc(dc, opnd_create_reg(DR_REG_A0),
+                               opnd_create_pc(pc + (3 << 12)));
+    test_instr_encoding_failure(dc, OP_auipc, pc + 4, instr);
```

Comment:
Add a comment here that we expect failure since the `target_pc - instr_encode_pc` has non-zero lower 12 bits.

Reply:
Done.

Reply:
Unfortunately, `instr_encode_to_copy` is asserting on reachability, thus causing CI to fail. We could use `instr_encode_check_reachability` but that's an unexported procedure.

Reply:
Perhaps we can export instr_encode_check_reachability. @derekbruening that does seem useful for tests and maybe for clients too?

Reply:
There have been a couple of recent instances where we wanted to turn off reachability while encoding: #6196 and code that wants to get an x86/aarch32 encoding template (via get_encoding_info()) without any final PC in mind.  We might do the same thing for instr categories for #6238.  Those are all inside DR but I guess the point is that yes it is useful to control reachability checks.  Wait, it's asserting?  For client or any other use you would think we'd want to either fail or ignore, not assert: asserting is really only for internal DR encoding of blocks indicating it messed up the type of branch and we want a clearer error than "failed to encode".  Probably we should change all calls from clients to not assert and instead fail; and possibly allow ignoring reachability too.

Reply:
> Wait, it's asserting? For client or any other use you would think we'd want to either fail or ignore, not assert: asserting is really only for internal DR encoding of blocks indicating it messed up the type of branch and we want a clearer error than "failed to encode".

Yes; our implementation of `instr_encode_arch` asserts for reachability when its `assert_reachable` arg is set to true, e.g. in `instr_encode_to_copy`.

> Probably we should change all calls from clients to not assert and instead fail; and possibly allow ignoring reachability too.

Here are the three instr encode APIs that use `instr_encode_arch`: https://github.com/DynamoRIO/dynamorio/blob/fff1147153dd388faa1b69893a4b0910e35d484d/core/ir/encode_shared.c#L116

Maybe we should export all three of the above APIs and let the client decide which one they want? Even though asserting is not ideal, that's the behaviour clients have been relying on till now.

Reply:
Sorry I know too little about DynamoRIO so I can't get involved in the discussion...

Reply:
@ksco The suggestion is to set assert_reachable to false in the implementation of instr_encode_to_copy, and note this change in release.dox as a minor compatibility change. With this change, failure in encoding would not cause a crash and we can still write tests that detect reachability issues.

Reply:
Done.